### ISSUE TITLE
LeRobotMixtureDataset repeats the same underlying samples across epochs with persistent_workers=True

### ISSUE THREAD

**Issue by Alioth7** [starVLA/starVLA#470]
## Summary
Critical sampling bug: with the default `persistent_workers=True`, StarVLA’s `LeRobotMixtureDataset` can replay the same underlying samples every epoch (no even shuffling), leaving a substantial portion of the dataset potentially unseen throughout training.

DataLoader采样严重bug：默认 `persistent_workers=True` 会让 StarVLA 的 `LeRobotMixtureDataset` 在每个 epoch 反复训练同一批样本（连顺序打乱也没有），导致大量数据在整个训练过程中可能从未被看到。

### Describe the bug

`LeRobotMixtureDataset` is documented to return **different samples every epoch in `mode="train"`**, using:

```python
safe_hash((self.epoch, index, self.seed))
```

inside `sample_step()`.

However, under the default VLA DataLoader configuration with:

```python
num_workers > 0
persistent_workers = True
shuffle = False
```

the worker processes keep the dataset copies created during the first DataLoader iteration. Although Accelerate updates the **main-process** dataset epoch on every new iterator, the persistent worker copies do not observe that update and continue using:

```python
self.epoch == 0
```

for the entire training run.

As a result, every DataLoader index continues to map to the same underlying:

```text
(dataset, trajectory, step)
```

triple across epochs.

Therefore, the random-with-replacement sample set generated during the first epoch becomes effectively frozen. Frames that were not selected in that first mapping may never be sampled during the rest of training, regardless of how many times the DataLoader is exhausted.

This affects VLA training through:

```text
get_vla_dataset
    -> LeRobotMixtureDataset
    -> DataLoader
    -> accelerator.prepare(...)
```

when multiprocessing with persistent workers is enabled.

**I am afraid that this is a significant bug.**

---

### Expected behavior

For training mode, every epoch should use a different value of `epoch` when evaluating:

```python
safe_hash((epoch, index, seed))
```

so that the same DataLoader index samples a different trajectory / step across epochs.

Worker processes executing `__getitem__()` must therefore observe the updated epoch value.

---

#### 1. Mixture sampling explicitly depends on `self.epoch`

`LeRobotMixtureDataset` contains:

```python
def set_epoch(self, epoch: int):
    self.epoch = epoch

def sample_step(self, index: int):
    seed = (
        index
        if self.mode != "train"
        else safe_hash((self.epoch, index, self.seed))
    )

    rng = np.random.default_rng(seed)

    dataset_index = rng.choice(
        len(self.datasets),
        p=self.dataset_sampling_weights,
    )

    trajectory_index = rng.choice(
        len(dataset.trajectory_ids),
        p=self.trajectory_sampling_weights[dataset_index],
    )

    trajectory_id = dataset.trajectory_ids[trajectory_index]

    base_index = rng.choice(
        dataset.trajectory_lengths[trajectory_index]
    )

    return dataset, trajectory_id, base_index
```

The class documentation also states:

```text
If "train", __getitem__ will return different samples every epoch
```

So changing `self.epoch` is required for the intended sampling behavior.

---

#### 2. Accelerate does update the dataset epoch in the main process

StarVLA uses:

```text
accelerate==1.5.2
```

and the VLA DataLoader is passed through:

```python
accelerator.prepare(...)
```

Accelerate's `DataLoaderShard.__iter__()` calls:

```python
self.set_epoch(self.iteration)
```

before creating the underlying DataLoader iterator.

Its `set_epoch()` supports custom datasets and can call:

```python
self.dataset.set_epoch(epoch)
```

Therefore, this is **not** simply a missing `dataset.set_epoch()` call in the main training process.

The main-process `LeRobotMixtureDataset.epoch` does advance.

---

#### 3. Persistent DataLoader workers do not see the updated epoch

The VLA DataLoader currently defaults to persistent workers when `num_workers > 0`:

```python
dataloader_kwargs = {
    "batch_size": ...,
    "collate_fn": collate_fn,
    "num_workers": num_workers,
    "pin_memory": ...,
    # shuffle=True
}

if num_workers > 0:
    dataloader_kwargs["persistent_workers"] = bool(
        vla_dataset_cfg.get("persistent_workers", True)
    )
```

On the `starVLA` branch, the same behavior was previously hard-coded:

```python
num_workers=16
persistent_workers=True
# shuffle=True
```

With PyTorch multiprocessing, workers hold their own dataset copies.

Conceptually:

```text
first iterator:

main dataset epoch = 0
        |
        +--> worker 0 dataset copy: epoch = 0
        +--> worker 1 dataset copy: epoch = 0
        +--> worker 2 dataset copy: epoch = 0
        +--> worker 3 dataset copy: epoch = 0
```

On the next iterator, Accelerate does:

```text
main dataset epoch = 1
```

but the persistent workers are not recreated:

```text
worker 0 dataset copy: epoch = 0
worker 1 dataset copy: epoch = 0
worker 2 dataset copy: epoch = 0
worker 3 dataset copy: epoch = 0
```

Since `__getitem__()` and `sample_step()` run in those workers, they continue evaluating:

```python
safe_hash((0, index, seed))
```

instead of:

```python
safe_hash((1, index, seed))
safe_hash((2, index, seed))
...
```

---

#### 4. DataLoader indices themselves are also not shuffled

VLA DataLoader shuffle is disabled:

```python
# shuffle=True
```

So each epoch iterates the same index domain.

For a given index `i`, the worker repeatedly computes:

```python
safe_hash((0, i, seed))
```

which deterministically produces the same:

```text
dataset_index
trajectory_id
base_index
```

every time that index is visited.

Therefore, the underlying sample mapping is frozen across epochs.

Note that stochastic image transforms may still make the final tensors differ slightly, but the underlying `(dataset, trajectory, step)` selection is unchanged.

---

### Impact

`LeRobotMixtureDataset` samples with replacement rather than constructing a permutation over all frames.

Therefore, a single epoch generally does not cover every frame.

Normally this is fine because changing `epoch` creates a new deterministic sample mapping on the next pass.

With persistent workers stuck at `epoch=0`, however, the same with-replacement sample set is replayed for the entire training run.

This means:

* some frames are sampled multiple times every epoch;
* some frames selected in epoch 0 are replayed throughout training;
* some frames not selected by the epoch-0 mapping may never appear at all;
* larger datasets in a mixture may remain significantly under-covered depending on mixture weights.

For example, in one 5-dataset equal-weight mixture:

```text
total source frames: ~186k
LeRobotMixtureDataset.__len__(): 268160
```

the frozen epoch mapping covered approximately:

```text
~73% unique source frames
```

meaning approximately:

```text
~27% of source frames were never sampled
```

during the entire training run if the mapping remained frozen.

The exact fraction depends on dataset sizes, trajectory lengths, weights, and seed, but the repeated mapping behavior is general.

---

### Minimal reproduction

A minimal reproduction does not require robot data.

For example, use a dataset whose `__getitem__()` returns the epoch value seen by the worker:

```python
class EpochDataset(torch.utils.data.Dataset):
    def __init__(self):
        self.epoch = 0

    def set_epoch(self, epoch):
        self.epoch = epoch

    def __len__(self):
        return 8

    def __getitem__(self, index):
        return self.epoch, index
```

With:

```python
loader = DataLoader(
    dataset,
    batch_size=4,
    num_workers=2,
    persistent_workers=True,
)
```

calling:

```python
dataset.set_epoch(0)
list(loader)

dataset.set_epoch(1)
list(loader)

dataset.set_epoch(2)
list(loader)
```

still results in workers observing epoch 0 after the first iterator.

With:

```python
persistent_workers=False
```

the workers are recreated and observe the new epoch values.

---

### Suggested fix

The epoch value used by `sample_step()` must be visible inside persistent worker processes.

Possible solutions:

#### Option 1: disable persistent workers

Use:

```python
persistent_workers=False
```

for the VLA DataLoader.

This is the simplest correctness fix because workers will be recreated after each DataLoader exhaustion and receive the updated dataset state.

The downside is worker startup overhead.

---

#### Option 2: store epoch in shared memory

Keep persistent workers, but replace the process-local integer:

```python
self.epoch
```

with worker-visible shared state, for example using `multiprocessing.Value` or an equivalent mechanism.

Conceptually:

```python
self._shared_epoch = multiprocessing.Value("q", 0)

def set_epoch(self, epoch):
    self._shared_epoch.value = epoch

def sample_step(self, index):
    epoch = self._shared_epoch.value

    seed = (
        index
        if self.mode != "train"
        else safe_hash((epoch, index, self.seed))
    )
```

This preserves persistent-worker performance while making epoch-dependent sampling correct.

---

#### Option 3: make epoch part of the sampler/index

Another robust design would avoid mutable dataset state entirely and encode the epoch or random seed into the indices sent to workers.

For example:

```text
(epoch, index)
```

could be passed to `__getitem__()`.

Then worker-local dataset state would no longer need synchronization.

---

### About `shuffle=True`

Re-enabling:

```python
shuffle=True
```

would reshuffle DataLoader indices and is likely desirable for training, but it is **not a complete fix**.

If worker-side `epoch` remains frozen at 0, then:

```text
index i -> same (dataset, trajectory, step)
```

still holds.

Shuffling only changes the order in which the frozen mappings are visited.

The epoch synchronization bug should therefore be fixed independently.

---

### Affected configurations

This issue affects configurations using:

```text
mode="train"
num_workers > 0
persistent_workers=True
```

with `LeRobotMixtureDataset`.

It does **not** reproduce in the same way when:

```text
num_workers=0
```

because sampling runs in the main process, or when:

```text
persistent_workers=False
```

because workers are recreated and receive the updated dataset state.

---

### Related

* #399 makes `num_workers`, `persistent_workers`, `pin_memory`, and `prefetch_factor` configurable, but does not synchronize epoch state across workers.
* #332 / #333 concern DataLoader throughput configuration and metrics and do not change mixture epoch sampling.

---

### Environment

* Branch: `starVLA_dev`
* Also reproducible in the `starVLA` branch behavior using persistent workers
* `accelerate==1.5.2`
* Dataset format and video backend are unrelated. In our project, we use LeRobot v3.
* Root cause is `LeRobotMixtureDataset` epoch-dependent sampling combined with persistent multiprocessing workers
